packages/core/src/routes/experience/classes/experience-interaction.ts

```diff
@@ -0,0 +1,165 @@
+    );
+
+    const { verificationRecords = [], profile, accountId, event } = result.data;
+
+    this.interactionEvent = event;
+    this.accountId = accountId; // TODO: @simeng-li replace with userId
+    this.profile = profile;
+
+    this.verificationRecords = new Set(
+      verificationRecords.map((record) => buildVerificationRecord(libraries, queries, record))
+    );
+  }
+
+  /** Set the interaction event for the current interaction */
+  public setInteractionEvent(event: InteractionEvent) {
+    // TODO: conflict event check (e.g. reset password session can't be used for sign in)
+    this.interactionEvent = event;
+  }
+
+  /** Set the verified accountId of the current interaction from  the verification record */
+  public identifyUser(verificationRecord: VerificationRecord) {
+    // Throws an 404 error if the user is not found by the given verification record
+    assertThat(
+      verificationRecord.verifiedUserId,
+      new RequestError({
+        code: 'user.user_not_exist',
+        status: 404,
+      })
+    );
+
+    this.accountId = verificationRecord.verifiedUserId;
+  }
+
+  /**
+   * Append a new verification record to the current interaction.
+   * @remark If a record with the same type already exists, it will be replaced.
+   */
+  public appendVerificationRecord(record: VerificationRecord) {
+    const { type } = record;
+
+    const existingRecord = this.getVerificationRecordByType(type);
+
+    if (existingRecord) {
+      this.verificationRecords.delete(existingRecord);
+    }
+
+    this.verificationRecords.add(record);
+  }
+
+  public getVerificationRecordById(verificationId: string) {
+    return this.verificationRecordsArray.find((record) => record.id === verificationId);
+  }
+
+  /** Save the current interaction result */
+  public async save() {
+    // The "mergeWithLastSubmission" will only merge current request's interaction results,
+    // manually merge with previous interaction results
+    // refer to: https://github.com/panva/node-oidc-provider/blob/c243bf6b6663c41ff3e75c09b95fb978eba87381/lib/actions/authorization/interactions.js#L106
+
+    const { provider } = this.tenant;
+    const details = await provider.interactionDetails(this.ctx.req, this.ctx.res);
+
+    await provider.interactionResult(
+      this.ctx.req,
+      this.ctx.res,
+      { ...details.result, ...this.toJson() },
+      { mergeWithLastSubmission: true }
+    );
+  }
+
+  /** Submit the current interaction result to the OIDC provider and clear the session */
+  public async submit() {
+    // TODO: refine the error code
+    assertThat(this.accountId, 'session.verification_session_not_found');
+
+    const { provider } = this.tenant;
+
+    const redirectTo = await provider.interactionResult(this.ctx.req, this.ctx.res, {
+      login: { accountId: this.accountId },
+    });
+
+    this.ctx.body = { redirectTo };
+  }
+
+  private get verificationRecordsArray() {
+    return Array.from(this.verificationRecords);
+  }
```

Comment:
no need for this? just `[...this.verificationRecords]` for `Set` or `[...this.verificationRecords.values()]` if you are using `Map`